### ISSUE TITLE
Thread synchronisation issue during cache access.

### ISSUE THREAD

**Issue by asandeep** [peterbe/premailer#225]
Looks like I got hit with a thread synchronisation issue in Premailer's cache. Here's my usecase:

I am parsing thousands of HTML documents every hour using Premailer. Now each document can have its own set of CSS defined inline. Apart from that, I have a fixed CSS file that I apply to every document that gets processed. To speed things up, I start multiple thread workers that parses the document in parallel.

The program runs fine till the time Premailer's cache gets completely filled, after which it starts removing least recently used items. Now since multiple threads are accessing same cache there are chances that a rule removed by one thread was being access by another. Below is the traceback of exception that is raised in this case:

`File "/home/ubuntu/Documents/setups/anaconda2/envs/sentieo/lib/python2.7/site-packages/premailer/premailer.py", line 442, in transform
   sel = _create_cssselector(selector)
 File "/home/ubuntu/Documents/setups/anaconda2/envs/sentieo/lib/python2.7/site-packages/cachetools/__init__.py", line 48, in wrapper
   cache[k] = v
 File "/home/ubuntu/Documents/setups/anaconda2/envs/sentieo/lib/python2.7/site-packages/cachetools/lru.py", line 21, in __setitem__
   cache_setitem(self, key, value)
 File "/home/ubuntu/Documents/setups/anaconda2/envs/sentieo/lib/python2.7/site-packages/cachetools/cache.py", line 52, in __setitem__
   self.popitem()
 File "/home/ubuntu/Documents/setups/anaconda2/envs/sentieo/lib/python2.7/site-packages/cachetools/lru.py", line 35, in popitem
   return (key, self.pop(key))
 File "/home/ubuntu/Documents/setups/anaconda2/envs/sentieo/lib/python2.7/site-packages/cachetools/abc.py", line 39, in pop
   raise KeyError(key)
KeyError: (u'.separator',)`

The cachetools documentation already states that Cache implementation classes are not thread-safe (See note section here: https://cachetools.readthedocs.io/en/latest/#cache-implementations). But it is possible to synchronise thread by passing a lock instance to cachetools memoizing decorators (https://cachetools.readthedocs.io/en/latest/#cachetools.cached).

For now I worked my way around this by increasing the cache size and hacking the cachetools.LRUCache implementation to return LFUCache instead (code below) which works in my case for now but this just delays the disaster and not the full-proof solution.

```python
def _hacked_lfu_cache(maxsize):
    """
    This function replaces the LFUCache constructor to instead return LFUCache
    instance.
    """
    return cachetools.LFUCache(1000)

# Premailer initializes its cache with LFUCache first time it is imported. This
# hack replaces the constructor with a mock function that returns LFUCache
# instance before Premailers does the import.
cachetools.LRUCache = _hacked_lfu_cache
```

Another problem that I see which might not be that common but I would like to highlight it anyway. Premailer stores rules in cache with CSS selector as key. Now in cases (Such as mine :P) where multiple unrelated documents are being processed using Premailer, there are high chances of selector collision i.e. a selector defined by one document might be used by another document with a completely different set of rules. In that case, Premailer will apply the rules that are available in its cache, resulting in incorrect rules being applied.

So, I propose below changes to fix above issues:

1. Implement thread safety for cache access. Can be implicit or explicit via flag or init parameter.
2. Provide a way to use a different cache implementation. Selecting any of the default implementations would do, and having option to provide own implementation would be great.
3. To fix last one, may be style rules can be stored with some Id that uniquely identifies the document the rule belongs to.

I hope the bug provides enough information. If now, please let me know. Happy to send PR to fix couple of them.

**Comment by peterbe** [comment]
That's awesome! Great work. 

By the way, why can't you use multiple processes instead? It has higher startup costs but remedies much of the thread safety problems.

**Comment by asandeep** [comment]
I actually do. So each document has hundreds of recepients. Each document is processed in a sub process of its own and then it spawns multiple threads to deliver emails to those recepients. Since, each email needs to be customised for receipient, I use premailer to apply the styling and thats where it causes the synchronization issue.

**Comment by peterbe** [comment]
Ok. Can you reproduce some of these problems in a unit test? If so you should be able to easily prove a solution in a pull request. 

I will probably not attempt to solve this myself, because I don't use premailer any more myself (not that I use anything else either). But I'm happy to help with the maintenance of a code review.

**Comment by asandeep** [comment]
Hey Peter,

Sorry it took me some time to respond back. 

I have raised a PR for points (1) and (2) above. I couldn't add tests for (1), as it is hard to make this race condition happen everytime without modifying cachetools code. Just FYI, I was able to reproduce it by adding a bit of delay in cachetools code. I can share the code if you want.

I was wrong about (3). After further diving into Premailer's code, I realized that the selector and rules are cached separately and So, won't be causing the collision. Sorry about that.

Here is the link to PR for your review: https://github.com/peterbe/premailer/pull/227
